### ISSUE TITLE
Issues with ssh-agent on my windows machine

### ISSUE THREAD

**Issue by tobiasehlert** [microsoft/vscode-remote-release#2743]
- VSCode Version: 1.44.0
- Local OS Version: Windows_NT x64 10.0.17763
- Remote OS Version: CentOS
- Remote Extension/Connection Type: SSH

```
[21:40:47.443] Log Level: 2
[21:40:47.452] remote-ssh-nightly@2020.4.9540
[21:40:47.452] win32 x64
[21:40:47.453] SSH Resolver called for "ssh-remote+7b22686f73744e616d65223a22536562626543656e744f53227d", attempt 1
[21:40:47.454] SSH Resolver called for host: XXXCentOS
[21:40:47.454] Setting up SSH remote "XXXCentOS"
[21:40:47.474] Using commit id "2aae1f26c72891c399f8604XXXXXXXXXXXXXXXXX" and quality "stable" for server
[21:40:47.475] Install and start server if needed
[21:40:48.984] Checking ssh with "ssh -V"
[21:40:49.034] > OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
[21:40:49.037] Running script with connection command: ssh -T -D 51416 XXXCentOS bash
[21:40:49.040] Terminal shell path: C:\Windows\System32\cmd.exe
[21:40:49.250] > 
[21:40:49.250] Got some output, clearing connection timeout
[21:40:49.588] > Enter passphrase for key 'C:\Users\myuseraccount\.ssh\theprivatekey.ppk':
[21:40:49.588] Detected passphrase message
[21:40:51.360] Got passphrase response
[21:40:51.362] "install" wrote data to terminal: "******************"
[21:40:51.404] > 
> 
[21:40:51.623] > mylinuxuser@198.XXX.XXX.XXX: Permission denied (publickey,gssapi-keyex,gssapi-with-mi
> c).
> The process tried to write to a nonexistent pipe.
> 
[21:40:51.871] "install" terminal command done
[21:40:51.871] User entered passphrase
[21:40:51.872] Passphrase was entered and ssh config entry was found - adding SSH key to agent
[21:40:51.873] Executing add-ssh-key:
powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1
[21:40:51.874] Terminal shell path: C:\Windows\System32\cmd.exe
[21:40:51.874] "add-ssh-key" wrote data to terminal: "powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1"
[21:40:52.031] > Microsoft Windows [Version 10.0.17763.1098]
> (c) 2018 Microsoft Corporation. All rights reserved.
> 
> C:\Users\myuseraccount>
> 
> 
[21:40:52.533] > ]0;  - powershell  -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1C:\Users\myuseraccount>(powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Te
> mp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0
> 
[21:40:52.807] > PS: adding key to ssh-agent
> ssh-add
> C:\Users\myuseraccount\.ssh\theprivatekey.ppk
> 
[21:40:52.843] > Error connecting to agent: No such file or directory
> 320bf9b4505b##5##
> 
[21:40:52.843] ssh-add failed
[21:40:52.843] Consider setting up the ssh agent: https://aka.ms/vscode-remote/troubleshooting#_setting-up-the-ssh-agent
[21:40:52.871] > ]0; 
[21:40:52.871] ssh-add failed
[21:40:52.871] Consider setting up the ssh agent: https://aka.ms/vscode-remote/troubleshooting#_setting-up-the-ssh-agent
[21:40:53.118] "add-ssh-key" terminal command done
[21:40:53.118] Install terminal quit with output: The process tried to write to a nonexistent pipe.
[21:40:53.118] Received install output: The process tried to write to a nonexistent pipe.
[21:40:53.119] Resolver error: The process tried to write to a nonexistent pipe
[21:40:53.123] ------
```

When running powershell, it seems to fail similar to vscode:
```
PS C:\Windows\system32> ssh-add -l
error fetching identities: communication with agent failed
```

And when checking for the SSH_AUTH_SOCK, it doens't exist on my workstation:
```
PS C:\Windows\system32> gci env:SSH_AUTH_SOCK
gci : Cannot find path 'SSH_AUTH_SOCK' because it does not exist.
```

Does this issue occur when you try this locally?: Yes
Does this issue occur when you try this locally and all extensions are disabled?: No

**Comment by roblourens** [comment]
It sounds like the agent is not running. That's not why you couldn't connect, the reason you couldn't connect is that you got a `Permission denied` error from the host.

**Comment by tobiasehlert** [comment]
Hi @roblourens,

I've checked and the agend it running under services, but I don't understand, why it's not possible to connect to it. The system environment list doesn't show any trail for it either (when running: gci env:SSH_AUTH_SOCK)

In addition, the SSH-key is working when I do use PuTTY and login to the same server :-(

Plus I've also tried it at a second computer, and it's also failing there.

One thing both computers have in common is that I have the WSL installed, but I've removed it from one of them and tried around a lot without problems.

Regards,
Tobias

**Comment by roblourens** [comment]
We don't support PuTTY, you may want to see the tips here: https://code.visualstudio.com/docs/remote/troubleshooting#_configuring-key-based-authentication

**Comment by tobiasehlert** [comment]
Hi @roblourens,

Well, what I was trying to point our with mentioning PuTTY was, that the key is working and I can login to the server. :)
And I've went through all the two sites you have mentioned and I didn't get it to work either.

I especially did check into this section:
https://code.visualstudio.com/docs/remote/troubleshooting#_setting-up-the-ssh-agent
_**Windows:**_
_To enable SSH Agent automatically on Windows, start a local Administrator PowerShell and run the following commands:_
### _Make sure you're running as an Administrator_
```
Set-Service ssh-agent -StartupType Automatic
Start-Service ssh-agent
Get-Service ssh-agent_
```
_Now the agent will be started automatically on login._

But.. then it still fails (with the block of log I pasted in my first submit of this issue.
I think that it is failing when trying to communicate with the ssh-agent:
```
[21:40:52.843] > Error connecting to agent: No such file or directory
> 320bf9b4505b##5##
```

In services in the Task Manager, the ssh-agent is listed as running:
![image](https://user-images.githubusercontent.com/4417850/79190921-b8f25380-7e25-11ea-8e71-eee71468039a.png)

What could the reason be for the client not to be able to communicate with the ssh-agent?

Regards,
Tobias

**Comment by tobiasehlert** [comment]
Somehow (magically) I had some more success now..

I switched to the Visual Studio Code - Insider version and use the remote-ssh nightly version as well. Also turned on logging to trace (and you see a fresh log at the bottom of this comment).

The ssh-agent seems to be running now:
```
PS C:\Users\myuseraccount> Get-Service ssh-agent
Status   Name               DisplayName
------   ----               -----------
Running  ssh-agent          OpenSSH Authentication Agent
```

And when checking for the keys that are loaded, I don't get an error right now
```
PS C:\Users\myuseraccount> ssh-add -l
4096 SHA256:rpPKRwR8XXXXXXXXXXXXXXXXXXXX/x9jEUXV9ZPC9So mylinuxuser@198.XXX.XXX.XXX (RSA)
```

And here comes the little bit longer log now:
```
[11:56:41.743] Log Level: 1
[11:56:41.749] remote-ssh-nightly@2020.4.9540
[11:56:41.749] win32 x64
[11:56:41.754] SSH Resolver called for "ssh-remote+198.XXX.XXX.XXX", attempt 1
[11:56:41.756] SSH Resolver called for host: 198.XXX.XXX.XXX
[11:56:41.756] Setting up SSH remote "198.XXX.XXX.XXX"
[11:56:41.799] Using commit id "c8826e4ff5246775ebee0fb932402de6af55ddc8" and quality "insider" for server
[11:56:41.803] Install and start server if needed
[11:56:43.346] Checking ssh with "ssh -V"
[11:56:43.426] > OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
[11:56:43.430] Running script with connection command: ssh -T -D 55959 198.XXX.XXX.XXX bash
[11:56:43.442] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[11:56:43.560] "install" terminal received data: "[2J[?25l[m[H

[H]0;C:\WINDOWS\System32\cmd.exe[?25h"
[11:56:43.560] Got some output, clearing connection timeout
[11:56:43.565] "install" terminal received data: "[?25l[80X[80C
[80X[80C
[...]
[80X[80C
[80X[80C[H[?25h"
[11:56:43.940] "install" terminal received data: "[?25lEnter passphrase for key 'C:\Users\myuseraccount\.ssh\theprivatekey.ppk': [?25h"
[11:56:43.940] Detected passphrase message
[11:56:46.408] Got passphrase response
[11:56:46.409] "install" wrote data to terminal: "*****"
[11:56:46.423] "install" terminal received data: "[?25l
[?25h"
[11:56:46.715] "install" terminal received data: "[?25lmylinuxuser@198.XXX.XXX.XXX: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).[77X[77C
[?25h"
[11:56:46.725] "install" terminal received data: "[?25lThe process tried to write to a nonexistent pipe.
[?25h"
[11:56:46.998] "install" terminal command done
[11:56:46.998] User entered passphrase
[11:56:47.000] Passphrase was entered and ssh config entry was found - adding SSH key to agent
[11:56:47.001] Executing add-ssh-key:
powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1
[11:56:47.002] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[11:56:47.004] "add-ssh-key" wrote data to terminal: "powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1"
[11:56:47.039] "add-ssh-key" terminal received data: "[2J[?25l[m[H

[H]0;C:\WINDOWS\System32\cmd.exe[?25h"
[11:56:47.046] "add-ssh-key" terminal received data: "[?25lMicrosoft Windows [Version 10.0.18362.720][10X[10C
(c) 2019 Microsoft Corporation. All rights reserved.
[52X[52C
C:\Users\myuseraccount>[36X[36C[4;17H[?25h"
[11:56:47.530] "add-ssh-key" terminal received data: "[?25l
C:\Users\myuseraccount>(powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0[29X[29C
[?25h"
[11:56:47.538] "add-ssh-key" terminal received data: "[?25l[HMicrosoft Windows [Version 10.0.18362.720][38X[38C
(c) 2019 Microsoft Corporation. All rights reserved.[28X[28C
[80X[80C
C:\Users\myuseraccount>(powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0[29X[29C
[80X[80C
[...]
[80X[80C
[80X[80C[6;1H]0;C:\WINDOWS\System32\cmd.exe - powershell  -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1[?25h"
[11:56:47.624] "add-ssh-key" terminal received data: "[?25l[HMicrosoft Windows [Version 10.0.18362.720][38X[38C
(c) 2019 Microsoft Corporation. All rights reserved.[28X[28C
[80X[80C
C:\Users\myuseraccount>(powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0[29X[29C
[80X[80C
[...]
[80X[80C
[80X[80C[6;1H[?25h"
[11:56:47.942] "add-ssh-key" terminal received data: "[?25lPS: adding key to ssh-agent[?25h"
[11:56:47.950] "add-ssh-key" terminal received data: "[?25l[HMicrosoft Windows [Version 10.0.18362.720][38X[38C
(c) 2019 Microsoft Corporation. All rights reserved.[28X[28C
[80X[80C
C:\Users\myuseraccount>(powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0[29X[29C
PS: adding key to ssh-agent[53X[53C
ssh-add[73X[73C
C:\Users\myuseraccount\.ssh\theprivatekey.ppk[49X[49C
[80X[80C
[...]
[80X[80C
[80X[80C[9;1H[?25h"
[11:56:47.971] "add-ssh-key" terminal received data: "[?25lEnter passphrase for C:\Users\myuseraccount\.ssh\theprivatekey.ppk: [?25h"
[11:56:47.971] Detected passphrase message
[11:56:47.972] Got passphrase response
[11:56:47.972] "add-ssh-key" wrote data to terminal: "*****"
[11:56:47.990] "add-ssh-key" terminal received data: "[?25l
[?25h"
[11:56:48.190] "add-ssh-key" terminal received data: "[?25lIdentity added: C:\Users\myuseraccount\.ssh\theprivatekey.ppk (mylinuxuser@198.XXX.XXX.XXX)[77X[77C
[?25h"
[11:56:48.198] "add-ssh-key" terminal received data: "[?25l[HMicrosoft Windows [Version 10.0.18362.720][38X[38C
(c) 2019 Microsoft Corporation. All rights reserved.[28X[28C
[80X[80C
C:\Users\myuseraccount>(powershell -ExecutionPolicy Bypass C:\Users\myuseraccount\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0[29X[29C
PS: adding key to ssh-agent[53X[53C
ssh-add[73X[73C
C:\Users\myuseraccount\.ssh\theprivatekey.ppk[49X[49C
Enter passphrase for C:\Users\myuseraccount\.ssh\theprivatekey.ppk:[27X[27C
Identity added: C:\Users\myuseraccount\.ssh\theprivatekey.ppk (mylinuxuser@198.XXX.XXX.XXX)[77X[77C
[80X[80C
[...]
[80X[80C
[80X[80C[12;1H[?25h"
[11:56:48.198] Detected passphrase message
[11:56:48.198] Got passphrase response
[11:56:48.198] "add-ssh-key" wrote data to terminal: "*****"
[11:56:48.229] "add-ssh-key" terminal received data: "[?25l]0;C:\WINDOWS\System32\cmd.exe[?25h"
[11:56:48.239] "add-ssh-key" terminal received data: "[?25l[?25h"
[11:56:48.512] "add-ssh-key" terminal command done
[11:56:48.513] Install terminal quit with output: The process tried to write to a nonexistent pipe.
[11:56:48.513] Received install output: The process tried to write to a nonexistent pipe.
[11:56:48.514] Resolver error: The process tried to write to a nonexistent pipe
[11:56:48.517] TELEMETRY: {"eventName":"resolver","properties":{"outcome":"failure","reason":"OfflineError","askedPw":"0","askedPassphrase":"1","asked2fa":"0","askedHostKey":"0","gotUnrecognizedPrompt":"0","remoteInConfigFile":"1"},"measures":{"resolveAttempts":1,"retries":1}}
[11:56:48.520] ------
```

Regards,
Tobias

**Comment by roblourens** [comment]
Still

> `Permission denied (publickey,gssapi-keyex,gssapi-with-mic)`

We run a command that's basically like

`type <file> | ssh <yourhost>`

You can write `uname` to some file, run the command above, and if you can get that to work, then vscode should work.

**Comment by tobiasehlert** [comment]
@roblourens, I succeeded finally! :)

What I did now:

1. Created a brand new key with type Ed25519 with puttygen.
2. Set a password
3. Saved the public and private key
4. Exported the key as "Export OpenSSH key"
5. Added the public key to ~/.ssh/authorized_keys
6. In normal CMD.exe I added the key with ssh-add
7. Run SSH command and got logged in
8. Tested to connect with VSCode Insider and got logged in

Anyhow.. issue is still closed, so nothing more to do here.
Thanks for being patient with me 👍
